### DISCUSSION TITLE
How should we think about `winget pin`

### DISCUSSION THREAD

**Discussion by denelon** [microsoft/winget-cli#2980] [PowerShell]
Pin-WinGetPackage isn't an approved verb

We also have it as an experimental feature, so we will look at the PowerShell idoms for experimental features, and make it correctly align to enable the feature in winget via the PowerShell - COM API interface.

I did get the suggestion of "Lock" on the PowerShell community call today.

**Comment by SeeminglyScience** [comment]
Just to throw out a couple more, one of these is probably what I'd do:

- `Lock-WinGetPackage`
- `Limit-WinGetPackage`
- `Set-WinGetPackageVersionPin`

**Reply by denelon** [reply]
I'll split these out for a 👍 style vote.

**Reply by joelst** [reply]
Lock and Limit are valid and correctly describe what is being done. Definitely extra points for being able use standardized verbs that we don't see often enough correctly. I'm just not sure PowerShellers will gravitate towards them.

**Comment by denelon** [comment]
Lock-WinGetPackage

**Reply by denelon** [reply]
I think I like this one. 

One caveat to note is the package might upgrade itself, other package managers, and the user could still upgrade without going through WinGet.

One of the concepts we thought about is some mechanism to display packages at a newer version than the PIN as a "pin violated" type of indication.

**Reply by ALIENQuake** [reply]
Even if Lock seems to fit nicely, the description of this verb is "Secures a resource" so the usage context is somehow related to 'security'. Currently, this verb [it's commonly used as so](https://github.com/search?q=language%3APowerShell+lock-++&type=code&l=PowerShell). Given the case, I lean towards [@joelst proposal](https://github.com/microsoft/winget-cli/discussions/2980#discussioncomment-4999700).

**Comment by denelon** [comment]
Limit-WinGetPackage

**Reply by jantari** [reply]
I believe this is the best choice. The description of the `Limit` verb fits best: "Applies constraints to a resource" because it most accurately describes what we're doing. `Lock` and `Unlock`, like @ALIENQuake said, are intended and mostly used in the security context. But more importantly, a "winget Package Version Pin" may not LOCK a package to one specific version, instead the pin should (either today or in the near future) allow for version CONSTRAINT expressions such as what [npm](https://docs.npmjs.com/cli/v8/configuring-npm/package-json#dependencies) or [terraform](https://developer.hashicorp.com/terraform/language/expressions/version-constraints) accept.

Then, when I "pin" (constrain) a package to for example `">= 1.2.0, < 2.0.0"` we really cannot say we're locking it, but we are most certainly limiting it.

Also, the `Limit-WinGetPackage` cmdlet could in the future be re-used for other kinds of limits or constraints we might want to configure for a package. For example we might want to limit the sources it should be updated from. This functionality could easily be incorporated through additional parameters:

```powershell
# fictional example, not real syntax
Limit-WinGetPackage -Package "my-funny-program" -Version ">= 1.2.0, < 2.0.0" -Sources "internal-enterprise-source-123"
```

`Lock` sounds better at first glance, in part I think because `Limit` is one of the rarely used verbs. But I hope we can keep `Limit` in the discussion for the reasons I outlined here, even if it's anti-democratic 😄

**Reply by denelon** [reply]
Limit may actually be a better verb. The discussion is intended to uncover things like this. 😊

Before we make any decision, we'll share the reasoning we use.

**Reply by cgerke** [reply]
On second thoughts Limit definitely comes across more logical. Discoverability probably not so important for this tool, most people using should be fairly advanced with pwsh.

**Reply by ThomasNieto** [reply]
The only problem with limit is what verb are you using to remove the limit? There is no paired verb with limit.

**Reply by denelon** [reply]
Good question @ThomasNieto. One approach would be logically setting the "limit" to none or unlimited. I guess the other approach would be to make a pin into a noun and use the other verbs.

**Reply by joelst** [reply]
Is this still an option open for discussion? 
My unbiased opinion 🙄 this [Add-WingetPackagePin, Remove-WingetPackagePin](https://github.com/microsoft/winget-cli/discussions/2980#discussioncomment-4999700) was a decent suggestion 😜 

I think it mirrors the `winget pin` syntax while still being in line with PowerShell standards.

It could work like this:

```PowerShell
# fictional example, not real syntax
Add-WinGetPackagePin -Package "my-funny-program" -Versions @(">= 1.2.0", "< 2.0.0") -Sources @("internal-enterprise-source-123")
Remove-WinGetPackagePin -Package "my-funny-program"
```

**Reply by ALIENQuake** [reply]
There is also a way to keep 'Limit':
```powershell
New-WinGetPackageLimit -Name "my-funny-program" -SemVersion ">= 1.2.0", "< 2.0.0"
Set-WinGetPackageLimit -Name "my-funny-program" -SemVersion ">= 1.2.1", "< 2.0.1" -Source "internal-enterprise-source-123"
Remove-WinGetPackageLimit -Name "my-funny-program"
```

**Reply by denelon** [reply]
* https://github.com/microsoft/winget-cli/issues/3234

We can certainly change the title and incorporate this suggestion.
Throw some comments on the issue and we'll see if we get pushback or consensus. 😊

**Reply by Umrhedar** [reply]
> * [Add "Limit-WinGetPackage" #3234](https://github.com/microsoft/winget-cli/issues/3234)
> 
> 
> We can certainly change the title and incorporate this suggestion. Throw some comments on the issue and we'll see if we get pushback or consensus. 😊

I like the way it is done with "apt" in debian/ubuntu like: "apt-mark hold <app-name>" so with winget and powershell I would suggest
winget --hold -Name/ID <app-name/-id> -Version <>=x.y.z
winget --unhold -Name/ID <app-name/-id> -Version <>=x.y.z
winget --list-hold

or like
Set-WinGetHold -add -Name/ID <app-name/-id> -Version <>=x.y.z
Set-WinGetHold -remove -Name/ID <app-name/-id> -Version <>=x.y.z
Get-WinGetHold

At the moment I just use a script with all App-IDs I want to upgrade instead of using --all because there is one userfull little bugger driving me crazy with the new version... On my linux systems the hold feature is my saviour in many situations until I have time to change something or a problem is fixed. It's a good feature when used with consideration. I'm looking forward to see/use that feature in winget.

**Comment by denelon** [comment]
Set-WinGetPackageVersionPin

**Reply by denelon** [reply]
This one might cause a bit of a headache with autocomplete.

**Comment by joelst** [comment]
I think it ok to use `New-` and `Remove-` or `Set-` and `Reset-` 

Since the cli command is `winget pin **add**` it really should be `Add-` and `Remove-` that way PowerShell is in line with the cli. This will make it easier to use. 

I like `Lock-` however, it is not used very often, so it may not be easy to discover, which is ironic since the standardized verbs are meant to make it easy to discover. 

The noun would be something like `WinGetPackagePin` 

My suggestion is the following: 

- `Add-WinGetPackagePin` - Add a pin to the specified package
- `Remove-WinGetPackagePin` - Remove the pin from the specified package
- `Get-WinGetPackagePin` - Return a list of all pins, or the pin for a specified package.

**Reply by cgerke** [reply]
@joelst suggestions seems most logical and easily discoverable for those new to powershell.

**Reply by SeeminglyScience** [reply]
I'd actually lean towards these tbh

**Comment by denelon** [comment]
* #3234
